src/main/scala/amba/apb/ToTL.scala

```diff
@@ -0,0 +1,83 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.amba.apb
+
+import Chisel._
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.MaskGen
+
+case class APBToTLNode()(implicit valName: ValName) extends MixedAdapterNode(APBImp, TLImp)(
+  dFn = { case APBMasterPortParameters(masters) =>
+    TLClientPortParameters(clients = masters.map { m =>
+      TLClientParameters(name = m.name, nodePath = m.nodePath)
+    })
+  },
+  uFn = { mp => APBSlavePortParameters(
+    slaves = mp.managers.map { m =>
+      APBSlaveParameters(
+        address       = m.address,
+        resources     = m.resources,
+        regionType    = m.regionType,
+        executable    = m.executable,
+        nodePath      = m.nodePath,
+        //TODO why and how should I do this adjust for APB?
+        // Also, shouldn't I use supportsPutPartial because of PSTRB?
+        supportsWrite = m.supportsPutFull,//adjust(m.supportsPutFull),
+        supportsRead  = m.supportsGet)},//adjust(m.supportsGet))},
+    beatBytes = mp.beatBytes)
+  }) {
+  //TODO: Shouldn't there be a requirement that TL side is wider than APB side?
+}
+
+class APBToTL()(implicit p: Parameters) extends LazyModule
+{
+  val node = APBToTLNode()
+
+  lazy val module = new LazyModuleImp(this) {
+    (node.in zip node.out) foreach { case ((in, edgeIn), (out, edgeOut)) =>
+      val beatBytes = edgeOut.manager.beatBytes
+
+      //-----------------------------------------------
+      // Flow Control
+      out.a.valid := in.psel && in.penable
```

Comment:
This is completely wrong. psel and penable stay high for the whole access phase. TileLink a valid indicates a transfer each time you assert it. You need a state machine like this:

val sent = RegInit(false.B)
out.a.valid := in.psel && !sent
when (out.a.fire()) { sent := true.B }
when (out.d.fire()) { sent := false.B }

Reply:
Ahhhh, the fact that penable stays high is the part I was missing from reading the spec. got it.

Reply:
Check out Figure 4-1 in the APB spec. It explains it well.